parts/linux/cloud-init/artifacts/cse_install.sh

```diff
@@ -229,20 +230,45 @@ installAzureCNI() {
 }
 
 extractKubeBinaries() {
-    K8S_VERSION=$1
-    KUBE_BINARY_URL=$2
+    local k8s_version="$1"
+    local kube_binary_url="$2"
 
     mkdir -p ${K8S_DOWNLOADS_DIR}
-    K8S_TGZ_TMP=${KUBE_BINARY_URL##*/}
-    retrycmd_get_tarball 120 5 "$K8S_DOWNLOADS_DIR/${K8S_TGZ_TMP}" ${KUBE_BINARY_URL} || exit $ERR_K8S_DOWNLOAD_TIMEOUT
-    tar --transform="s|.*|&-${K8S_VERSION}|" --show-transformed-names -xzvf "$K8S_DOWNLOADS_DIR/${K8S_TGZ_TMP}" \
+    local k8s_tgz_tmp=${kube_binary_url##*/}
+    retrycmd_get_tarball 120 5 "$K8S_DOWNLOADS_DIR/${k8s_tgz_tmp}" ${kube_binary_url} || exit $ERR_K8S_DOWNLOAD_TIMEOUT
+    tar --transform="s|.*|&-${k8s_version}|" --show-transformed-names -xzvf "$K8S_DOWNLOADS_DIR/${k8s_tgz_tmp}" \
         --strip-components=3 -C /usr/local/bin kubernetes/node/bin/kubelet kubernetes/node/bin/kubectl
-    rm -f "$K8S_DOWNLOADS_DIR/${K8S_TGZ_TMP}"
+    rm -f "$K8S_DOWNLOADS_DIR/${k8s_tgz_tmp}"
 }
 
-installKubeletKubectlAndKubeProxy() {
+extractPrivateKubeBinaries() {
+    local k8s_version="$1"
+    local kube_binary_url="$2"
+
+    local k8s_tgz_tmp=${kube_binary_url##*/}
+    local cached_pkg="${K8S_CACHE_DIR}/${k8s_tgz_tmp}"
+
+    if [[ -f "${cached_pkg}" ]]; then
+        echo "cached package ${cached_pkg} is found, will use that"
+    else
+        echo "cached package ${cached_pkg} not found"
+        return 1
+    fi
+
+    # remove the current kubelet binaries before extracting new binaries from the cached package
+    rm -rf /usr/local/bin/kubelet-* /usr/local/bin/kubectl-*
```

Comment:
These binaries look like 
/usr/local/bin/kubelet-<version> / /usr/local/bin/kubelet-1.28.0 for example.
At build time, all the configured versions are downloaded from the public URL, at run-time, the required version is kept and all the other ones are deleted (in the caller - mv /usr/local/bin/kubelet-<version> /usr/local/bin/kubelet).
In case of private and custom url, we delete all the current version to overwrite with the cached one at run-time.